hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionRegistryRpcStubHolder.java

```diff
@@ -0,0 +1,212 @@
+
+  private final Configuration conf;
+
+  // used for getting cluster id
+  private final Configuration noAuthConf;
+
+  private final User user;
+
+  private final RpcControllerFactory rpcControllerFactory;
+
+  private final Set<ServerName> bootstrapNodes;
+
+  private final int rpcTimeoutMs;
+
+  private volatile ImmutableMap<ServerName, ClientMetaService.Interface> addr2Stub;
+
+  private volatile RpcClient rpcClient;
+
+  private CompletableFuture<ImmutableMap<ServerName, ClientMetaService.Interface>> addr2StubFuture;
+
+  ConnectionRegistryRpcStubHolder(Configuration conf, User user,
+    RpcControllerFactory rpcControllerFactory, Set<ServerName> bootstrapNodes) {
+    this.conf = conf;
+    if (User.isHBaseSecurityEnabled(conf)) {
+      this.noAuthConf = new Configuration(conf);
+      this.noAuthConf.set(User.HBASE_SECURITY_CONF_KEY, "simple");
+    } else {
+      this.noAuthConf = conf;
+    }
+    this.user = user;
+    this.rpcControllerFactory = rpcControllerFactory;
+    this.bootstrapNodes = Collections.unmodifiableSet(bootstrapNodes);
+    this.rpcTimeoutMs = (int) Math.min(Integer.MAX_VALUE,
+      conf.getLong(HConstants.HBASE_RPC_TIMEOUT_KEY, HConstants.DEFAULT_HBASE_RPC_TIMEOUT));
+  }
+
+  private ImmutableMap<ServerName, ClientMetaService.Interface> createStubs(RpcClient rpcClient,
+    Collection<ServerName> addrs) {
+    LOG.debug("Going to use new servers to create stubs: {}", addrs);
+    Preconditions.checkNotNull(addrs);
+    ImmutableMap.Builder<ServerName, ClientMetaService.Interface> builder =
+      ImmutableMap.builderWithExpectedSize(addrs.size());
+    for (ServerName masterAddr : addrs) {
+      builder.put(masterAddr,
+        ClientMetaService.newStub(rpcClient.createRpcChannel(masterAddr, user, rpcTimeoutMs)));
+    }
+    return builder.build();
+  }
+
+  private void createStubsAndComplete(String clusterId,
+    CompletableFuture<ImmutableMap<ServerName, ClientMetaService.Interface>> future) {
+    RpcClient c = RpcClientFactory.createClient(conf, clusterId);
+    ImmutableMap<ServerName, ClientMetaService.Interface> m = createStubs(c, bootstrapNodes);
+    rpcClient = c;
+    addr2Stub = m;
+    future.complete(m);
+  }
+
+  private void createStubs(List<ServerName> bootstrapServers, int index,
+    CompletableFuture<ImmutableMap<ServerName, ClientMetaService.Interface>> future) {
+    // use null cluster id here as we do not know the cluster id yet, we will fetch it through this
+    // rpc client and use it to create the final rpc client
+    RpcClient getConnectionRegistryRpcClient = RpcClientFactory.createClient(noAuthConf, null);
+    // user and rpcTimeout are both not important here, as we will not actually send any rpc calls
+    // out, only a preamble connection header, but if we pass null as user, there will be NPE in
+    // some code paths...
+    RpcChannel channel =
+      getConnectionRegistryRpcClient.createRpcChannel(bootstrapServers.get(index), user, 0);
```

Comment:
Even if the rpcTimeout is not used, I think it's preferable to pass a real value in. We have a real value in this class, let's pass it in? Maybe something changes down the line, I don't like passing a 0 value for timeout

Reply:
This is intentional, I do not want to let developers think the timeout will take effect, as not everyone will look at the comment unless there is something unusual...

Reply:
After refactoring, I close the rpc client in the newly introduce ClusterIdFetcher class, after finishing all the rpc requests.